### ISSUE TITLE
Can't access `ELEVENTY_ENV` data

### ISSUE THREAD

**Issue by tempertemper** [11ty/eleventy#1886]
I'm trying to set the `ELEVENTY_ENV` in my site data file, following the [Exposing environment variables](https://www.11ty.dev/docs/data-js/#example-exposing-environment-variables) it's not returning anything.

Here's my set up:

I have a site.js file that lives in src/site/_data/

```
module.exports = {
  "title": "tempertemper",
  "author": {
    "name": "Martin Underhill",
    "twitter": "@tempertemper",
  },
  // The rest of the global variables
  "env": process.env.ELEVENTY_ENV
}
```

All of the data in there is working fine, for example `{{ site.title }}` outputs `tempertemper`, and `{{ site.author.twitter }}` output `@tempertemper`. `{{ site.env }}` doesn't output anything. It'd doesn't *break* anything either, which could be significant 🤷‍♂️

I've tried an `if` statement, like the Eleventy docs show:

```
{% if site.env == "production" %}
    Production!
{% else %}
    Development!
{% endif %}
```

At first I thought the `if` was needed, and it wasn't able to be output to the page with `{{ site.env }}`, but it turned out it was just failing on the match and falling back to the `else` statement: looking for a match on "development" and switching the if/else output just output "Production!".

I'm running:
- node version 16.4.2
- Eleventy 0.12.1
- macOS 11.4

I've tried [exposing the Eleventy environment as a filter](https://github.com/11ty/eleventy/issues/861#issuecomment-770463298) too, but no joy that way either.

I also tried setting the `site.url` data [using a ternary, like this Issue](https://github.com/11ty/eleventy/issues/233#issuecomment-422691098), but it's not finding 'development' and uses the `else` (`:`) value instead.

The closest issue to mine looks like [Environment function is not working #540](https://github.com/11ty/eleventy/issues/540), but that doesn't look like it was resolved.

Thanks for taking a look 💙 In case a poke around my code is useful, [here's my repo](https://github.com/tempertemper/tempertemper.net).

**Comment by zachleat** [comment]
What's the file name of your data file?

**Comment by pdehaan** [comment]
Where/how are you setting the `ELEVENTY_ENV` environment variable?

**Comment by pdehaan** [comment]
My understanding is that `ELEVENTY_ENV` and `NODE_ENV` aren't defined by default, you'd have to set those yourself (via the command line, or an .env file or whatever).

```js
// src/site/_data/site.js
module.exports = {
  title: "tempertemper",
  author: {
    name: "Martin",
  },
  // Use the `ELEVENTY_ENV` if defined, default to "development" otherwise.
  env: process.env.ELEVENTY_ENV || "development",
};
```

```njk
Global: {{ site.env }}
Filter: {{ "ELEVENTY_ENV" | env }}
```

Similarly, here's the `env` filter I quickly threw together (which falls back to some predefined default if the specified ENV key isn't set):

```js
// .eleventy.js (snippet)
eleventyConfig.addFilter("env", (key, def="NOT DEFINED") => process.env[key] || def);
```

Now if I run <kbd>eleventy</kbd> from some npm script, it will output the following:

```yaml
Global: development
Filter: NOT DEFINED
```

So the site.js used the fallback value of "development" for the `process.env.ELEVENTY_ENV`, but our `env` helper used the default fallback value of "NOT DEFINED", as expected. If you want to pass a default to the filter version, this should do the trick:

```njk
Filter: {{ "ELEVENTY_ENV" | env("development") }}
```

Next, here's a snippet of my package.json `scripts` block:

```js
  "scripts": {
    "build": "eleventy",
    "build:dev": "ELEVENTY_ENV=development eleventy",
    "build:prod": "ELEVENTY_ENV=production eleventy",
    "test": "echo \"Error: no test specified\" && exit 1"
  },
```

If I run <kbd>npm run build</kbd>, the `ELEVENTY_ENV` variable doesn't get set, and it will use whatever fallback values you may or may not have set up.

Now if I run <kbd>npm run build:dev</kbd>, it will set the `ELEVENTY_ENV` variable to "development" (at least on macOS, it's naturally more complicated if you want this to work on macOS and Windows in a compatible way) and rebuild my site, and my output will be:

```yaml
Global: development
Filter: development
```

Similarly, running <kbd>npm run build:prod</kbd> will give us the following:


```yaml
Global: production
Filter: production
```

---

**Bonus tip:** If you want to dump all your env vars in your site (which you most certainly don't for various security related reasons), this was useful for some debugging:

```js
  // requires `npm i sort-json`
  eleventyConfig.addFilter("sort_json", require("sort-json"));
  eleventyConfig.addFilter("inspect", require("util").inspect);
```

Then in my "env.njk" file, I could use the following:

```njk
{{ env | sort_json | inspect | safe }}
```

I won't paste the output, since I lazily have some GitHub API tokens and stuff in my env vars, but it's pretty interesting to see all the npm env vars that npm sets when running via <kbd>npm run [script]</kbd>. Plus a good reminder that variables set through environment vars seem to always come in as strings, so depending on what you're trying to access, you might need to convert into Booleans or Numbers as needed:

```yaml
  npm_config_fetch_retry_mintimeout: '10000',
  npm_config_force: '',
  npm_config_format_package_lock: 'true',
```

**Comment by tempertemper** [comment]
> Where/how are you setting the `ELEVENTY_ENV` environment variable?

I'm ashamed to admit that I misinterpreted the documentation and assumed Eleventy came with its own environment variables. It did cross my mind as to how it would know I was running a local or production environment, but as with a lot of things development, I assumed magic (maybe if Browsersync was running it would be in dev mode, otherwise production 🤷‍♂️). Maybe because the example environment variable had an eleventy namespace?

Anyway, I added an environment variable to my .env file:

```
ELEVENTY_ENV=development
```

Registered it in my .eleventy.js file:

```js
require('dotenv').config()
const { ELEVENTY_ENV } = process.env
```

Added it to my site data file:

```js
module.exports = {
  title: "tempertemper",
  author: {
    name: "Martin",
  },
  "env": process.env.ELEVENTY_ENV
};
```

And it works nicely in my templates:

```js
{%- if site.env == 'development' %}
    This is the development environment
{%- else %}
    This is not the development environment (production, staging, etc.)
{%- endif %}
```

So, first of all, thank you *very* much for the prompt @pdehaan 🏆

On top of that, I really appreciate all of the extra tips:

- The fallback value is a really nice touch and I'll be using that
- Passing environment variables via commands is a great idea – it's not private, so it doesn't need to live in the gitignored .env file like an API key might, so it can happily live in my package.json file
- The bonus tip has been duly lodged in my memory and will certainly come in useful at some point

Peter, you're awesome 🙇‍♂️
